src/Tools/BuildValidator/CompilationOptionsReader.cs

```diff
@@ -6,60 +6,265 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.IO;
+using System.IO.Compression;
 using System.Linq;
+using System.Reflection;
 using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Cci;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Newtonsoft.Json;
 
 namespace BuildValidator
 {
+    internal readonly struct SourceFileInfo
+    {
+        internal string SourceFilePath { get; }
+        internal SourceHashAlgorithm HashAlgorithm { get; }
+        internal byte[] Hash { get; }
+        internal SourceText? EmbeddedText { get; }
+
+        internal SourceFileInfo(
+            string sourceFilePath,
+            SourceHashAlgorithm hashAlgorithm,
+            byte[] hash,
+            SourceText? embeddedText)
+        {
+            SourceFilePath = sourceFilePath;
+            HashAlgorithm = hashAlgorithm;
+            Hash = hash;
+            EmbeddedText = embeddedText;
+        }
+    }
+
     internal class CompilationOptionsReader
     {
+        public static readonly Guid HashAlgorithmSha1 = unchecked(new Guid((int)0xff1816ec, (short)0xaa5e, 0x4d10, 0x87, 0xf7, 0x6f, 0x49, 0x63, 0x83, 0x34, 0x60));
```

Comment:
Is this copied from somewhere?

Reply:
@ryzngard Yep.

https://github.com/dotnet/roslyn/blob/afd10305a37c0ffb2cfb2c2d8446154c68cfa87a/src/Compilers/Core/Portable/Debugging/SourceHashAlgorithms.cs#L16-L19

Reply:
fwiw if these need to be the same, a note about the source is a good comment :)

Reply:
Should I just rely on the definitions in SourceHashAlgorithms.cs and delete the copies of them in CompilationOptionsReader.cs? That requires the IVT to the compiler, but maybe it is reasonable to assume that IVT will be around indefinitely. What do you think @jaredpar?

Reply:
Do they need to be the same, or is there some other reason these magical numbers work?

Reply:
Rather than IVT can we just use a shared source approach?

Reply:
The HashAlgorithm is a real part of the portable PDB spec. https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/specs/PortablePdb-Metadata.md#document-table-0x30

I'll look into sharing source when I get the chance. Perhaps looking at the prior art in Workspaces, etc. will make it clear how I should do this.

Reply:
oh, just noting that it's part of the Portable PDB spec would be a great way to demystify this. I don't think sharing code would be needed